### ISSUE TITLE
<% a %> in knitr source code

### ISSUE THREAD

**Issue by gaborcsardi** [jimhester/lintr#185]
lintr fails on knits with this:

```
Error in grep(pattern$chunk.begin, lines, perl = TRUE) :
  invalid 'pattern' argument
```

To reproduce:

```r
cat("'<% a %>'\n", file = tmp <- tempfile());
lintr::lint(tmp)
#> Error in grep(pattern$chunk.begin, lines, perl = TRUE) :
#>   invalid 'pattern' argument
```

**Comment by jimhester** [comment]
This looks to be because in knitr the [brew pattern has only an inline pattern](https://github.com/yihui/knitr/blob/f5e2d9e12f6981ac7144986f1aa2c014ff28e09e/R/pattern.R#L17), not a chunk start or end.

FWIW In devel lintr you can now pass strings directly to `lintr::lint()` if you end them with a newline.

``` r
 lintr::lint("'<% a %>'\n")
#> <text>:1:1: style: Only use double-quotes.
#> '<% a %>'
#> ^~~~~~~~~
#> <text>:2:1: style: Trailing blank lines are superfluous.
#> 
#> ^
```

**Comment by gaborcsardi** [comment]
Thanks! 

Btw. `knitr` is nasty because its files are full of the knitr patterns, obviously. see #127 

Do you need to use those patterns for parsing? I haven't looked much, but in reneral it would be great to parse without them, or at least not everything.

> FWIW In devel lintr you can now pass strings directly to lintr::lint() if you end them with a newline.

But then how do I lint my files whose name ends with a newline? :P

**Comment by jimhester** [comment]
after https://github.com/jimhester/lintr/commit/72f198585026cb347fcf1678b505b0e12b4d8303, https://github.com/jimhester/lintr/commit/7fe7e9093c7e9f41eebe600d89a98375d68f8f40 knitr is now fully lintable and returns 5,569 lints with the default linters :).

Maybe not necessary for everything, it seemed easier to do it this way rather than have a file extension whitelist or something.

**Comment by gaborcsardi** [comment]
I don't mind the lints, just the errors....

Anyway, maybe I'll take a look, I'll pick up goodpractice soon again, and will do some lintr related stuff.
